loopy/target/c/codegen/expression.py

```diff
@@ -544,6 +562,36 @@ def map_group_hw_index(self, expr, type_context):
     def map_local_hw_index(self, expr, type_context):
         raise LoopyError("plain C does not have local hw axes")
 
+    def map_nan(self, expr, type_context):
+        from loopy.types import NumpyType
+        if expr.data_type is None:
+            if type_context == "f":
+                return p.Variable("NAN")
+            elif type_context == "d":
+                registry = self.codegen_state.ast_builder.target.get_dtype_registry()
+                lpy_type = NumpyType(np.dtype(np.float32))
+                cast = var("(%s)" % registry.dtype_to_ctype(lpy_type))
+                return cast(p.Variable("NAN"))
+            else:
+                raise NotImplementedError("lowering NaN with type context"
+                                          f" '{type_context}'.")
+        else:
+            if isinstance(expr.data_type(float("nan")), np.float32):
```

Comment:
Not wanting to hold up the merge here, but why is this not `if expr.data_type == np.float32`?

FWIW:
```
>>> np.dtype(np.float32) == np.float32
True
```
(Similar below.)

Reply:
expr.data_type is a bit more general (cf https://documen.tician.de/pymbolic/primitives.html?highlight=nan#pymbolic.primitives.NaN.data_type) i.e. can be a callable and not just `np.float32`.

Reply:
Ah, thanks!